README.md
# BSS-Response-Geant4
the calculation of a traditional  Bonner Sphere Spectrometer using Geant4


History
$Id: History 71661 2013-06-20 08:26:18Z gcosmo $

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                    Hadr03 History file
                    -------------------
This file should be used by the G4 example coordinator to briefly
summarize all major modifications introduced in the code and keep
track of all tags.

     ----------------------------------------------------------
     * Reverse chronological order (last date on top), please *
     ----------------------------------------------------------

19-06-13 mma (exhadr03-V09-06-19)
- SteppingVerbose: use G4Step::GetSecondaryInCurrentStep()
     
14-06-13 mma (exhadr03-V09-06-18)
- SteppingAction: use G4Step::GetSecondaryInCurrentStep()

05-06-13 I. Hrivnacova (exhadr03-V09-06-17)
- applied coding guidelines (adding base class initialization in constructors, 
  removed tabulations, fixed long lines)
     
28-05-13 mma (exhadr03-V09-06-16)
- cosmetic in SteppingVerbose
     
21-04-13 mma (exhadr03-V09-06-15)
- cleanup all macros for processName
     
24-03-13 mma (exhadr03-V09-06-14)
- polish coding conventions
     
16-03-13 mma (exhadr03-V09-06-13)
- restaure G4HadronPhysicsINCLXX    
- G4 coding conventions

13-03-13 V.Ivanchenko (exhadr03-V09-06-12)
- RunAction - change printout from "warning XXXX" to "NOTE:"
    because CDASH interpretation of "warning" is test failer 

12-03-13 V.Ivanchenko (exhadr03-V09-06-11)
- DetectorConstruction - improved material definition
- HistoManager - do not delete singleton
- PhsyicsList - new name for INCLXX physics constructor 
     
10-03-13 mma (exhadr03-V09-06-10)
- PhysicsList: remove G4HadronPhysicsQGSP_INCLXX
    
10-03-13 mma (exhadr03-V09-06-09)
- modify readme
     
09-03-13 mma (exhadr03-V09-06-08)
- add envHadronic.csh to control environment variables
     
06-03-13 mma (exhadr03-V09-06-07)
- RunAction - modify printing of cross section per atom
     
01-03-13 mma (exhadr03-V09-06-06)
- DetectorConstruction - suppress natural abundance flag
    
22-02-13 V.Ivanchenko (exhadr03-V09-06-05)
- DetectorConstruction - set natural abundances flag "false" for 
    elements built from one isotope

01-02-13 V.Ivanchenko (exhadr03-V09-06-04)
- PhysicsList - adopted migration to new physics_lists directory 
    structure and change of class names
     
01-02-13 mma (exhadr03-V09-06-03)
- DetectorMessenger: change /buildMatIsotope to /setIsotopeMat
      
31-01-13 mma (exhadr03-V09-06-02)
- DetectorMessenger: add command /buildMatIsotope 
- RunAction: add material in GetCrossSectionPerAtom()
     
08-01-13 mma (exhadr03-V09-06-01)
- DetectorConstruction: add function MaterialWithSingleIsotope()
     
04-12-12 mma (exhadr03-V09-06-00)
- PhysicsList: QGSP_BERT_HP replaced by FTFP_BERT_HP
- RunAction: compute crossSection per atom
     
23-11-12 mma (exhadr03-V09-05-15)
- SteppingAction : add more protection for neutronHP
     
21-11-12 mma (exhadr03-V09-05-14)
- PhysicsList: use QBBC instead of QGSP_BERT_HP
     
28-10-12 mma (exhadr03-V09-05-11)
- update README and .README

17-10-12 Ivana Hrivnacova (exhadr03-V09-05-10)
- Adding missing explicit use of system of units 
 
12-10-12 Andrea Dotti (exhadr03-V09-05-09)
- Adding explicit use of system of units

06-10-12 mma (exhadr03-V09-05-08)
- SteppingAction, RunAction : print number of gamma
- use G4Analysis
    
30-09-12 mma (exhadr03-V09-05-07)
- SteppingAction : print number of particles in nuclear reaction (not gamma)
     
18-08-12 mma (exhadr03-V09-05-06)
- SteppingAction : use GetTargetIsotope()
     
01-08-12 mma(exhadr03-V09-05-05) 
- RunAction : use GetCrossSectionPerVolume()
     
20-07-12 mma(exhadr03-V09-05-04) 
- GammaPhyics : rm crossSections
- add histo Pbalance
     
27-06-12 mma (exhadr03-V09-05-03)
- add .README
     
21-06-12 mma (exhadr03-V09-05-02)
- GammaPhyics : add crossSections
     
19-06-12 mma (exhadr03-V09-05-01)
- suppress local HadronPhysics and IonPhysics
     
15-06-12 mma (exhadr03-V09-05-00)
- rename Hadr14
     
13-06-12 Michel Maire (exhadr14-V09-05-00)
- created from TestEm14 and rdecay01


src/Run.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
// $Id: Run.cc 66536 2012-12-19 14:32:36Z ihrivnac $
//
/// \file Run.cc
/// \brief Implementation of the Run class

#include "Run.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

Run::Run()
: G4Run()
{
 for(G4int i=0;i<800;i++)
   CounterSpectrum[i]=0;
} 

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

Run::~Run()
{} 
 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void Run::Merge(const G4Run* run)
{
  const Run* localRun = static_cast<const Run*>(run);
  for(G4int i=0;i<800;i++)
     CounterSpectrum[i]+=localRun->CounterSpectrum[i];
  G4Run::Merge(run); 
} 

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
void Run::AddCounterSpectrum(G4int i)
{
  CounterSpectrum[i]++;
}
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......




src/EventAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
// $Id: EventAction.cc 75117 2013-10-28 09:38:37Z gcosmo $
//
/// \file EventAction.cc
/// \brief Implementation of the EventAction class

#include "EventAction.hh"
#include "Run.hh"

#include "G4Event.hh"
#include "G4RunManager.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

EventAction::EventAction()
: G4UserEventAction()
{
} 

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

EventAction::~EventAction()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void EventAction::BeginOfEventAction(const G4Event*)
{   
  fEdep=0.;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void EventAction::EndOfEventAction(const G4Event*)
{   
  // accumulate statistics in Run
  Run* run 
    = static_cast<Run*>(
        G4RunManager::GetRunManager()->GetNonConstCurrentRun());
  if(fEdep!=0)
  {
  G4int i=(int)(fEdep/0.001);
  if(i<800)
  run->AddCounterSpectrum(i);
 // else
  //G4cout<<"Warning   "<<fEdep<<G4endl;
  }
 // if(fEdep!=0)
 // G4cout<<fEdep<<G4endl;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......


src/PhysicsList.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
/// \file hadronic/Hadr03/src/PhysicsList.cc
/// \brief Implementation of the PhysicsList class
//
// $Id: PhysicsList.cc 70268 2013-05-28 14:17:50Z maire $

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "PhysicsList.hh"

#include "G4SystemOfUnits.hh"
#include "G4UnitsTable.hh"

#include "NeutronHPphysics.hh"

// particles

#include "G4BosonConstructor.hh"
#include "G4LeptonConstructor.hh"
#include "G4MesonConstructor.hh"
#include "G4BosonConstructor.hh"
#include "G4BaryonConstructor.hh"
#include "G4IonConstructor.hh"
#include "G4ShortLivedConstructor.hh"
#include "G4EmStandardPhysics_option3.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

PhysicsList::PhysicsList()
:G4VModularPhysicsList()
{
  SetVerboseLevel(1);
  
  //add new units
  //
  new G4UnitDefinition( "millielectronVolt", "meV", "Energy", 1.e-3*eV);   
  new G4UnitDefinition( "mm2/g",  "mm2/g", "Surface/Mass", mm2/g);
  new G4UnitDefinition( "um2/mg", "um2/mg","Surface/Mass", um*um/mg);  
    
  // Neutron Physics
  RegisterPhysics( new NeutronHPphysics("neutronHP")); 
  RegisterPhysics( new G4EmStandardPhysics_option3());  
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

PhysicsList::~PhysicsList()
{ }

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void PhysicsList::ConstructParticle()
{
  G4BosonConstructor  pBosonConstructor;
  pBosonConstructor.ConstructParticle();

  G4LeptonConstructor pLeptonConstructor;
  pLeptonConstructor.ConstructParticle();

  G4MesonConstructor pMesonConstructor;
  pMesonConstructor.ConstructParticle();

  G4BaryonConstructor pBaryonConstructor;
  pBaryonConstructor.ConstructParticle();

  G4IonConstructor pIonConstructor;
  pIonConstructor.ConstructParticle();

  G4ShortLivedConstructor pShortLivedConstructor;
  pShortLivedConstructor.ConstructParticle();  
}
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void PhysicsList::SetCuts()
{
  SetCutValue(0*mm, "proton");
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......


src/SteppingAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
// $Id: SteppingAction.cc 74483 2013-10-09 13:37:06Z gcosmo $
//
/// \file SteppingAction.cc
/// \brief Implementation of the SteppingAction class

#include "SteppingAction.hh"
#include "EventAction.hh"
#include "DetectorConstruction.hh"

#include "Run.hh"
#include "G4Step.hh"
#include "G4Event.hh"
#include "G4RunManager.hh"
#include "G4LogicalVolume.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

SteppingAction::SteppingAction(EventAction* eventAction)
: G4UserSteppingAction(),
  fEventAction(eventAction)
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

SteppingAction::~SteppingAction()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void SteppingAction::UserSteppingAction(const G4Step* step)
{
/*ParticleName=step->GetTrack()->GetDefinition()->GetParticleName();
if(ParticleName=="triton")
{
i=step->GetTrack()->GetCurrentStepNumber();
if(i==1)
{
Run* run = static_cast<Run*>(G4RunManager::GetRunManager()->GetNonConstCurrentRun());
VolumeName=step->GetTrack()->GetVolume()->GetName();
if(VolumeName=="He3")
run->AddCounter();
//G4double Energy=step->GetPreStepPoint()->GetKineticEnergy();
//G4cout<<"triton:  "<<Energy<<G4endl;
}
}
*/
/*else if(ParticleName=="proton")
{
i=step->GetTrack()->GetCurrentStepNumber();
VolumeName=step->GetTrack()->GetVolume()->GetName();
if(i==1&&VolumeName=="He3")
{
G4double Energy=step->GetPreStepPoint()->GetKineticEnergy();
G4cout<<"proton:  "<<Energy<<G4endl;
}
}*/

VolumeName=step->GetTrack()->GetVolume()->GetName();
if(VolumeName=="He3")
{
EnergyDespoit=step->GetTotalEnergyDeposit();
fEventAction->AddEnergyDespoit(EnergyDespoit);
}

}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......



src/ActionInitialization.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
// $Id: ActionInitialization.cc 68058 2013-03-13 14:47:43Z gcosmo $
//
/// \file ActionInitialization.cc
/// \brief Implementation of the ActionInitialization class

#include "ActionInitialization.hh"
#include "RunAction.hh"
#include "PrimaryGeneratorAction.hh"
#include "EventAction.hh"
#include "SteppingAction.hh"
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
ActionInitialization::ActionInitialization()
 : G4VUserActionInitialization()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

ActionInitialization::~ActionInitialization()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void ActionInitialization::BuildForMaster() const
{
  SetUserAction(new RunAction);
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void ActionInitialization::Build() const
{
  SetUserAction(new PrimaryGeneratorAction);
  SetUserAction(new RunAction);

  EventAction* eventAction = new EventAction;
  SetUserAction(eventAction);

  SetUserAction(new SteppingAction(eventAction));
}  

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......


src/PrimaryGeneratorAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
/// \file hadronic/Hadr03/src/PrimaryGeneratorAction.cc
/// \brief Implementation of the PrimaryGeneratorAction class
//
// $Id: PrimaryGeneratorAction.cc 70759 2013-06-05 12:26:43Z gcosmo $
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "PrimaryGeneratorAction.hh"

#include "G4Event.hh"
#include "G4GeneralParticleSource.hh"
#include "G4ParticleTable.hh"
#include "G4ParticleDefinition.hh"
#include "globals.hh"
#include "Randomize.hh"
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

PrimaryGeneratorAction::PrimaryGeneratorAction()
:G4VUserPrimaryGeneratorAction(),
 particleGun(0)                                              
{
  particleGun  = new G4GeneralParticleSource();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

PrimaryGeneratorAction::~PrimaryGeneratorAction()
{
  delete particleGun;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void PrimaryGeneratorAction::GeneratePrimaries(G4Event* anEvent)
{
 //G4cout<<"     "<<G4UniformRand()<<G4endl;
 particleGun->GeneratePrimaryVertex(anEvent);
}
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......



Hadr03.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
/// \file hadronic/Hadr03/Hadr03.cc
/// \brief Main program of the hadronic/Hadr03 example
//
//
// $Id: TestEm1.cc,v 1.16 2010-04-06 11:11:24 maire Exp $
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo...... 

#include "G4RunManager.hh"
#include "G4UImanager.hh"
#include "Randomize.hh"

#include "DetectorConstruction.hh"
#include "PhysicsList.hh"
#include "ActionInitialization.hh"
#include "G4VUserPhysicsList.hh"
#include "G4PhysListFactory.hh"
#include "G4VModularPhysicsList.hh"
#include "G4MTRunManager.hh"
#include "G4GenericPhysicsList.hh"
#ifdef G4VIS_USE
 #include "G4VisExecutive.hh"
#endif

#ifdef G4UI_USE
#include "G4UIExecutive.hh"
#endif

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
 
int main(int argc,char** argv) {
 
  //choose the Random engine
 G4Random::setTheEngine(new CLHEP::RanecuEngine);
//  CLHEP::HepRandom::setTheSeed(9876);
    
  // Construct the default run manager
  G4MTRunManager* runManager = new G4MTRunManager;
  runManager->SetNumberOfThreads(24);
//  G4RunManager * runManager = new G4RunManager;

  // set mandatory initialization classes
  DetectorConstruction* det=new DetectorConstruction();
  runManager->SetUserInitialization(det);

 /* std::vector<G4String>* MyConstr = new std::vector<G4String>;

      MyConstr->push_back("G4EmStandardPhysics");
      MyConstr->push_back("G4EmExtraPhysics");
      MyConstr->push_back("G4DecayPhysics");
      MyConstr->push_back("G4HadronElasticPhysics");
      MyConstr->push_back("G4HadronPhysicsQGSP_BERT_HP");
     // MyConstr->push_back("G4HadronPhysicsQGSP_BERT");
      MyConstr->push_back("G4StoppingPhysics");
      MyConstr->push_back("G4IonPhysics");
      MyConstr->push_back("G4NeutronTrackingCut");

     G4VModularPhysicsList*  phys = new G4GenericPhysicsList(MyConstr);

  runManager->SetUserInitialization(phys);*/

  // Physics list
     PhysicsList* phys = new PhysicsList;
  runManager->SetUserInitialization(phys);
/* G4PhysListFactory factory;
  G4String physName = "QGSP_BIC";
  G4VModularPhysicsList* phys = factory.GetReferencePhysList(physName);
  runManager->SetUserInitialization(phys);*/

 // G4PhysListFactory factory;
 // G4VModularPhysicsList* phys=factory.GetReferencePhysList("QGSP_BERT_HP");
  //G4VModularPhysicsList* phys=factory.GetReferencePhysList("QGSP_BIC_HP");
 // runManager->SetUserInitialization(phys);

  // User action initialization
 runManager->SetUserInitialization(new ActionInitialization());     
 runManager->Initialize();
  
#ifdef G4VIS_USE
  // Initialize visualization
 G4VisManager* visManager = new G4VisExecutive;
  // G4VisExecutive can take a verbosity argument - see /vis/verbose guidance.
  // G4VisManager* visManager = new G4VisExecutive("Quiet");
 visManager->Initialize();
#endif

  // Get the pointer to the User Interface manager
  G4UImanager* UImanager = G4UImanager::GetUIpointer();

  if (argc!=1) {
    // batch mode
    G4String command = "/control/execute ";
    G4String fileName = argv[1];
    UImanager->ApplyCommand(command+fileName);
  }
  else {
    // interactive mode : define UI session
#ifdef G4UI_USE
    G4UIExecutive* ui = new G4UIExecutive(argc, argv);
#ifdef G4VIS_USE
    UImanager->ApplyCommand("/control/execute energy1.mac"); 
#else
   // UImanager->ApplyCommand("/control/execute init.mac"); 
#endif
   // UImanager->ApplyCommand("/control/execute energy1.mac"); 
    ui->SessionStart();
    delete ui;
#endif
  }

  // Job termination
  // Free the store: user actions, physics_list and detector_description are
  // owned and deleted by the run manager, so they should not be deleted 
  // in the main() program !
  
#ifdef G4VIS_USE
  delete visManager;
#endif
  delete runManager;

  return 0;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo...... 
